### ISSUE TITLE
Fog of War: Loader data becomes undefined on navigation

### ISSUE THREAD

**Issue by ziimakc** [remix-run/remix#9799]
### Reproduction

Download: https://github.com/ziimakc/remix-routes-err
Run:
- `pnpm i`
- `pnpm dev`
- open `/` root page
- click `open` button
- click `error` button
- see an error

---

When navigating from `/user/1/max` to `user/1/edit` following error is shown:
> Cannot read properties of null (reading 'x')

because in `user.$id.$slug` data that we log `console.log(data.x)` from loader becomes `undefined`.

Note1: If we try to navigate from `/user/1/max` to `user/xxx` error will not happen
Note2: do not load `user/1/edit` page on directly because for some reason error will be gone , do a hard refresh.

### System Info

```shell
System:
    OS: Linux 6.8 Ubuntu 24.04 LTS 24.04 LTS (Noble Numbat)
    CPU: (8) x64 Intel(R) Core(TM) i5-9300H CPU @ 2.40GHz
    Memory: 9.64 GB / 15.46 GB
    Container: Yes
    Shell: 5.2.21 - /bin/bash
  Binaries:
    Node: 20.15.1 - /usr/bin/node
    npm: 10.7.0 - /usr/bin/npm
    pnpm: 9.4.0 - ~/.local/share/pnpm/pnpm
  Browsers:
    Chrome: 126.0.6478.182
  npmPackages:
    @remix-run/dev: 2.10.3 => 2.10.3 
    @remix-run/node: 2.10.3 => 2.10.3 
    @remix-run/react: 2.10.3 => 2.10.3 
    @remix-run/serve: 2.10.3 => 2.10.3 
    vite: 5.1.0 => 5.1.0
```

### Used Package Manager

pnpm

### Expected Behavior

There should be no error when navigating from `/user/1/max` to `user/1/edit`

### Actual Behavior

When navigating from `/user/1/max` to `user/1/edit` following error is shown:
> Cannot read properties of null (reading 'x')

**Comment by ziimakc** [comment]
@brophdawg11 this is an edit button that is an svg icon for which `onClick` action is easy to use/disable based on permissions. 

Maybe I can somehow use `Link` as a wrapper and fix css, but is not very convenient and this is not the only place where navigate is used. Basically this makes `navigate` dangerous to use. Will it be possible to disable `fogOfWar` in next version?

**Comment by brophdawg11** [comment]
Unsure - that's why it's still unstable 😉

**Comment by brophdawg11** [comment]
This is resolved by #9860 and will be available in the next release (2.11.2).  I confirmed the fix on the above repro using `2.11.2-pre.0` 👍
